atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/atomic/mcas/MarkAndCasConsensusForgettingStore.java

```diff
@@ -0,0 +1,210 @@
+
+        List<CasRequestBatch> pendingBatchedRequests = KeyedStream.stream(pendingUpdates.stream()
+                        .collect(Collectors.groupingBy(
+                                elem -> ByteBuffer.wrap(elem.argument().cell().getRowName()))))
+                .map((row, requests) -> new CasRequestBatch(tableRef, row, requests))
+                .values()
+                .collect(Collectors.toList());
+
+        // at this point, results for requests that will never be retried are already populated.
+        // There is one request per row
+        while (!pendingBatchedRequests.isEmpty()) {
+            pendingBatchedRequests = pendingBatchedRequests.stream()
+                    .filter(req -> !serveMcasRequest(kvs, req))
+                    .collect(Collectors.toList());
+        }
+    }
+
+    private static boolean serveMcasRequest(KeyValueService kvs, CasRequestBatch casRequestBatch) {
+        try {
+            kvs.multiCheckAndSet(casRequestBatch.getMcasRequest());
+            // The above operation is atomic
+            casRequestBatch.setSuccessForAllRequests();
+        } catch (MultiCheckAndSetException e) {
+            casRequestBatch.processBatchWithException(MarkAndCasConsensusForgettingStore::shouldRetry, e);
+        }
+        return casRequestBatch.isBatchServed();
+    }
+
+    // we only want to retry the requests where the actual matches the expected.
+    private static boolean shouldRetry(BatchElement<CasRequest, Void> req, MultiCheckAndSetException e) {
+        CasRequest casRequest = req.argument();
+        Cell cell = casRequest.cell();
+
+        if (!e.getActualValues().containsKey(cell)) {
+            return false;
+        }
+
+        return casRequest.expected().equals(ByteBuffer.wrap(e.getActualValues().get(cell)));
+    }
+
+    // Every request with a lower rank will never be tried. Requests for touch will fail with
+    // `CheckAndSetException` and those for atomic updates will fail with `KeyAlreadyExistsException`.
+    private static List<BatchElement<CasRequest, Void>> filterOptimalUpdatePerCell(
+            List<BatchElement<CasRequest, Void>> batch) {
+
+        Map<Cell, List<BatchElement<CasRequest, Void>>> partitionedElems = batch.stream()
+                .collect(Collectors.groupingBy(elem -> elem.argument().cell()));
+
+        ImmutableList.Builder<BatchElement<CasRequest, Void>> requestsToProcess = ImmutableList.builder();
+
+        for (Map.Entry<Cell, List<BatchElement<CasRequest, Void>>> requestedUpdatesForCell :
+                partitionedElems.entrySet()) {
+            List<BatchElement<CasRequest, Void>> sortedPendingRequests = requestedUpdatesForCell.getValue().stream()
+                    .sorted(Comparator.comparing(elem -> elem.argument().rank()))
+                    .collect(Collectors.toList());
+            if (!sortedPendingRequests.isEmpty()) {
+                requestsToProcess.add(sortedPendingRequests.remove(0));
```

Comment:
nit: removing the first request is O(n), we can instead just skip it in the next method